Add MIMXRT685 FlexSPI XIP support.

saurabh-nxp opened the pull request:
Driver APIs are tested using sample applications:
- samples/drivers/flash_shell 
- samples/subsys/fs/littlefs

These changes have dependency on PR #34086. So that MPU is setup to allow execution of driver code relocated to SRAM.

saurabh-nxp committed:
boards: arm: Enable FlexSPI driver on mimxrt685_evk.

Enable FlexSPI NOR flash driver with XIP build options.
Configure FlexSPI pins, update board documentation.
Add FlexSPI NOR flash dts node for mimxrt685_evk.

Enable flash storage by adding partition to dts file.
Using last 1MB for storage, and reserving 63MB for code.

Signed-off-by: Saurabh Jagdhane <saurabh.jagdhane@nxp.com>

saurabh-nxp committed:
drivers: flash: Enable FlexSPI XIP on mimxrt685_evk.

Move LUT to driver.
Update CMake to include SoC specific driver.
Fix mimxrt685_evk LUT header spacing.

Signed-off-by: Saurabh Jagdhane <saurabh.jagdhane@nxp.com>

MaureenHelm commented on boards/arm/mimxrt685_evk/flash_flexspi_nor_lut.h:
```diff
@@ -0,0 +1,66 @@
+
+#include "flash_mcux_flexspi_nor_lut_index.h"
+
+static const uint32_t flash_flexspi_nor_lut[][4] = {
```
Can this LUT be used for other octal flashes or is it specific to MX25UM51345G?

saurabh-nxp replied:
It is specific to MX25UM51345G. JEDEC standard is not valid for octal mode.

saurabh-nxp replied:
Not just memory specific but to do XIP or to trigger a read command sequence from LUT, dummy cycles are required. And dummy cycles requires as per memory model's data sheet varies as per frequency at which flash is operated. So along with memory model LUT is also platform/board specific.

MaureenHelm commented on boards/arm/mimxrt685_evk/flash_flexspi_nor_lut.h:
```diff
@@ -0,0 +1,66 @@
+	},
+
+	[READ_STATUS_REG] = {
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_SDR, kFLEXSPI_8PAD, 0x05,
+				kFLEXSPI_Command_SDR, kFLEXSPI_8PAD, 0xFA),
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_RADDR_SDR, kFLEXSPI_8PAD, 0x20,
+				kFLEXSPI_Command_DUMMY_SDR, kFLEXSPI_8PAD, 0x14),
+		FLEXSPI_LUT_SEQ(kFLEXSPI_Command_READ_SDR, kFLEXSPI_8PAD, 0x04,
+				kFLEXSPI_Command_STOP, kFLEXSPI_1PAD, 0x0),
```
Please fix the whitespace to follow the convention in the existing LUT

saurabh-nxp replied:
Done.

MaureenHelm commented on boards/arm/mimxrt1064_evk/flash_flexspi_nor_lut.h:
```diff
@@ -0,0 +1,89 @@
+
+#include "flash_mcux_flexspi_nor_lut_index.h"
+
+static const uint32_t flash_flexspi_nor_lut[][4] = {
```
I don't like defining the LUT at the board level because it will have to be duplicated for every other board that uses the same flash. I'd prefer instead to keep the LUT defined in `drivers/flash` because that allows reuse by both in-tree and out-of-tree boards. I think we should either 1) conditionalize the LUT in `drivers/flash/flash_mcux_flexspi_nor.c`, or 2) introduce a new `drivers/flash/flash_mcux_flexspi_ospi.c` or `drivers/flash/flash_mcux_flexspi_mx25um51345g.c`. I'm leaning toward option 2.

saurabh-nxp replied:
So `drivers/flash/flash_mcux_flexspi_mx25um51345g.c` would be required since LUT is memory specific. If not duplicating LUT header, is duplicating source code `flash_mcux_<memoryName>.c`  acceptable for planned device support in future?

MaureenHelm replied:
Yes

saurabh-nxp replied:
Ok, will do. Please take  a note that devices to be supported in future with same memory model may not be able to use this LUT. Because LUT, specifically dummy cycles used for READ sequence would change based on frequency at which flash is operated. Also, this LUT is valid only for Octal mode but in case of SPI/DPI/QPI modes with same memory model would require an update to LUT.

saurabh-nxp replied:
Done, moved LUT to driver.

pdgendt commented on boards/arm/mimxrt685_evk/pinmux.c:
```diff
@@ -278,6 +278,239 @@ static int mimxrt685_evk_pinmux_init(const struct device *dev)
+		IOPCTL_PIO_INV_DI);
+/* PORT2 PIN23 (coords: P5) is configured as FLEXSPI0B_DATA7 */
+IOPCTL_PinMuxSet(IOPCTL, 2U, 23U, port2_pin23_config);
+#endif
```
It kind of hurts my eyes, to see the same property and comment so many times, it is consistent with the code above though.

Maybe simplify to (untested & copy-pasted)
```c
const uint32_t common_pin_config = (
		/* Disable pull-up / pull-down function */
		IOPCTL_PIO_PUPD_DI |
		/* Enable pull-down function */
		IOPCTL_PIO_PULLDOWN_EN |
		/* Enables input buffer function */
		IOPCTL_PIO_INBUF_EN |
		/* Normal mode */
		IOPCTL_PIO_SLEW_RATE_NORMAL |
		/* Full drive enable */
		IOPCTL_PIO_FULLDRIVE_EN |
		/* Analog mux is disabled */
		IOPCTL_PIO_ANAMUX_DI |
		/* Pseudo Output Drain is disabled */
		IOPCTL_PIO_PSEDRAIN_DI |
		/* Input function is not inverted */
		IOPCTL_PIO_INV_DI);

/* PORT1 PIN11 (coords: L2) is configured as FLEXSPI0B_DATA0 */
IOPCTL_PinMuxSet(IOPCTL, 1U, 11U, common_pin_config | IOPCTL_PIO_FUNC6);
/* PORT1 PIN12 (coords: M2) is configured as FLEXSPI0B_DATA1 */
IOPCTL_PinMuxSet(IOPCTL, 1U, 12U, common_pin_config | IOPCTL_PIO_FUNC6);

/* etc */
```

But ignore me, if this is a policy :)

saurabh-nxp replied:
I just tried to keep it consistent with NXP's SDK code.

MaureenHelm replied:
I agree it's a bit overkill, but it is consistent with the rest of the file. This will soon go away anyway when we move the i.MX RT family to device tree pinmuxing.

pdgendt commented on drivers/memc/memc_mcux_flexspi.c:
```diff
@@ -140,6 +140,8 @@ static int memc_flexspi_init(const struct device *dev)
 #elif defined(CONFIG_XIP) && defined(CONFIG_CODE_FLEXSPI2)
 #define MEMC_FLEXSPI_CFG_XIP(node_id) DT_SAME_NODE(node_id, DT_NODELABEL(flexspi2))
+#elif defined(CONFIG_SOC_SERIES_IMX_RT6XX)
+#define MEMC_FLEXSPI_CFG_XIP(node_id) true
```
It solely depends on the soc type here, but XIP is enabled in the board defconfig.

```suggestion
#define MEMC_FLEXSPI_CFG_XIP(node_id) IS_ENABLED(CONFIG_FLASH_MCUX_FLEXSPI_XIP)
```
Or
```suggestion
#define MEMC_FLEXSPI_CFG_XIP(node_id) IS_ENABLED(CONFIG_XIP)
```

saurabh-nxp replied:
Done. Updated to `#define MEMC_FLEXSPI_CFG_XIP(node_id) IS_ENABLED(CONFIG_XIP)`

pdgendt commented on boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts:
```diff
@@ -143,6 +143,19 @@ arduino_spi: &flexcomm5 {
+		status = "okay";
+		jedec-id = [c2 81 3a];
+	};
+};
```
Move the relevant `reg` part to the flash device properties instead of the flash controller.

Similar to [PR](https://github.com/zephyrproject-rtos/zephyr/pull/34428)

MaureenHelm replied:
Let's get agreement on the direction in that PR first, before we change this one to follow.

mmahadevan108 commented on boards/arm/mimxrt685_evk/Kconfig.defconfig:
```diff
@@ -14,6 +14,15 @@ config XTAL_SYS_CLK_HZ
+	default y
+
+config FLASH_MCUX_FLEXSPI_XIP_MEM
+	default "SRAM"
```
Use `config FLASH_MCUX_FLEXSPI_XIP_MEM_TARGET` instead of `config FLASH_MCUX_FLEXSPI_XIP_MEM`. Else `#define CONFIG_FLASH_MCUX_FLEXSPI_XIP_MEM_ITCM 1` is set

saurabh-nxp replied:
Done.

saurabh-nxp committed:
drivers: memc: Update driver to support XIP on mimxrt685_evk.

Enable XIP based on SoC specific option.

Signed-off-by: Saurabh Jagdhane <saurabh.jagdhane@nxp.com>

mmahadevan108 commented:
Please update the https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/mimxrt685_evk/doc/index.rst#supported-features and https://github.com/zephyrproject-rtos/zephyr/blob/master/boards/arm/mimxrt685_evk/doc/index.rst#connections-and-ios sections of the document

mmahadevan108 commented:
Can you take a look at PR https://github.com/zephyrproject-rtos/zephyr/pull/34789 and add a partition section to the dts

saurabh-nxp committed:
samples: fs: Enable littlefs sample for mimxrt685_evk.

Add mimxrt685_evk_cm33 board to the platform_allow list
for the littlefs sample application.

Signed-off-by: Saurabh Jagdhane <saurabh.jagdhane@nxp.com>

mmahadevan108 commented on boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts:
```diff
@@ -176,6 +176,18 @@ i2s1: &flexcomm3 {
+			label = "storage";
+			reg = <0x03f00000 DT_SIZE_M(1)>;
+		};
+	};
```
This commit can be squashed with the earlier one that changes board dts file

saurabh-nxp replied:
After adding storage_partition to dts still facing an issue to get littlefs running. If able to get it to work will squash commits.

MaureenHelm replied:
The sample is failing because flash_area is trying to use the flexspi device with the flash API, when it should use the mx25um51345g device instead. You can fix this by moving the partitions section inside the mx25um51345g node.

``` Booting Zephyr OS build zephyr-v2.5.0-3120-g39d7242197e9  ***
Area 0 at 0x3f00000 on MX25UM51345G for 1048576 bytes
/lfs mount: 0
/lfs: bsize = 16 ; frsize = 4096 ; blocks = 256 ; bfree = 254
/lfs/boot_count stat: 0
        fn 'boot_count' siz 4
/lfs/boot_count read count 1: 4
/lfs/boot_count seek start: 0
/lfs/boot_count write new boot count 2: 4
/lfs/boot_count close: 0
/lfs opendir: 0
  F 4 boot_count
End of files
/lfs unmount: 0
[00:16:13.131,023] <inf> littlefs: LittleFS version 2.2, disk version 2.0
[00:16:13.133,602] <inf> littlefs: FS at MX25UM51345G:0x3f00000 is 256 0x1000-byte blocks with 512 e
[00:16:13.133,605] <inf> littlefs: sizes: rd 16 ; pr 16 ; ca 64 ; la 32
[00:16:13.133,876] <inf> littlefs: /lfs mounted
[00:16:13.161,600] <inf> littlefs: /lfs unmounted
```

saurabh-nxp replied:
Thank you! Done.

MaureenHelm commented on drivers/flash/CMakeLists.txt:
```diff
@@ -18,11 +18,19 @@ zephyr_library_sources_ifdef(CONFIG_SOC_FLASH_NIOS2_QSPI soc_flash_nios2_qspi.c)
 zephyr_library_sources_ifdef(CONFIG_SOC_FLASH_RV32M1 soc_flash_rv32m1.c)
 zephyr_library_sources_ifdef(CONFIG_FLASH_STM32_QSPI flash_stm32_qspi.c)
-zephyr_library_sources_ifdef(CONFIG_FLASH_MCUX_FLEXSPI_NOR flash_mcux_flexspi_nor.c)
+if(CONFIG_SOC_SERIES_IMX_RT6XX)
+	zephyr_library_sources_ifdef(CONFIG_FLASH_MCUX_FLEXSPI_NOR flash_mcux_flexspi_mx25um51345g.c)
+else()
+	zephyr_library_sources_ifdef(CONFIG_FLASH_MCUX_FLEXSPI_NOR flash_mcux_flexspi_nor.c)
+endif()
```
Please add a new config to `Kconfig.mcux` to make this:
```
zephyr_library_sources_ifdef(CONFIG_FLASH_MCUX_FLEXSPI_MX25UM51345G flash_mcux_flexspi_mx25um51345g.c)
```

And have the board enable this new config instead of CONFIG_FLASH_MCUX_FLEXSPI_NOR

That will allow you to remove the SoC-specific conditional here.

saurabh-nxp replied:
If I do that, I would have to add another config. based driver files inclusion in same CMakeLists.txt. Let me know if something as follows is ok to be added. Because this would duplicate existing `CONFIG_FLASH_MCUX_FLEXSPI_NOR` based option in some way.
`zephyr_include_directories_ifdef(
  CONFIG_FLASH_MCUX_FLEXSPI_MX25UM51345G
  ${ZEPHYR_BASE}/drivers/memc
)`

saurabh-nxp replied:
Changes are pushed. Please check.

MaureenHelm commented on drivers/flash/CMakeLists.txt:
```diff
@@ -50,7 +58,7 @@ zephyr_include_directories_ifdef(
 	CONFIG_SOC_FLASH_NRF_RADIO_SYNC_TICKER
 	${ZEPHYR_BASE}/subsys/bluetooth
 	${ZEPHYR_BASE}/subsys/bluetooth/controller/ll_sw/nordic
-	)
+)
```
please revert this unrelated change

saurabh-nxp replied:
Done.

MaureenHelm commented on drivers/flash/flash_mcux_flexspi_mx25um51345g.c:
```diff
@@ -0,0 +1,542 @@
+	return 0;
+}
+
+static int flash_flexspi_nor_enable_quad_mode(const struct device *dev)
```
replace `quad` with `octal`?

saurabh-nxp replied:
Done.

MaureenHelm commented on drivers/flash/CMakeLists.txt:
```diff
@@ -18,11 +18,16 @@ zephyr_library_sources_ifdef(CONFIG_SOC_FLASH_NIOS2_QSPI soc_flash_nios2_qspi.c)
 
 if(CONFIG_FLASH_MCUX_FLEXSPI_XIP)
-  zephyr_code_relocate(flash_mcux_flexspi_nor.c ${CONFIG_FLASH_MCUX_FLEXSPI_XIP_MEM}_TEXT)
+  if(CONFIG_SOC_SERIES_IMX_RT6XX)
```
Replace `CONFIG_SOC_SERIES_IMX_RT6XX` with `CONFIG_FLASH_MCUX_FLEXSPI_MX25UM51345G`

saurabh-nxp replied:
Done.

MaureenHelm commented on drivers/flash/Kconfig.mcux:
```diff
@@ -34,10 +34,17 @@ config FLASH_MCUX_FLEXSPI_NOR
 	select MEMC_MCUX_FLEXSPI
 
+config FLASH_MCUX_FLEXSPI_MX25UM51345G
+	bool "MCUX FlexSPI NOR driver"
```
Update this string to "MCUX FlexSPI MX25UM51345G driver"

saurabh-nxp replied:
Done.

MaureenHelm commented on boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts:
```diff
@@ -165,6 +165,19 @@ i2s1: &flexcomm3 {
+&flexspi {
+	reg = <0x50134000 0x4000>, <0x18000000 DT_SIZE_M(64)>;
+	mx25um51345g: mx25um51345g@0 {
+		compatible = "nxp,imx-flexspi-nor";
```
the compatible should be "nxp,imx-flexspi-mx25um51345g"

saurabh-nxp replied:
Done.

MaureenHelm commented on drivers/flash/flash_mcux_flexspi_mx25um51345g.c:
```diff
@@ -0,0 +1,550 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT	nxp_imx_flexspi_nor
```
this will become `nxp_imx_flexspi_mx25um51345g` after updating the dts compatible

saurabh-nxp replied:
Done.

saurabh-nxp commented:
I noticed some documentation and code compliance violation in CI workflows. Latest force-push addresses those only.

saurabh-nxp commented on boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts:
```diff
@@ -170,6 +170,30 @@ i2s1: &flexcomm3 {
 
+&flexspi {
+	reg = <0x50134000 0x4000>, <0x18000000 DT_SIZE_M(64)>;
+	mx25um51345g: mx25um51345g@0 {
```
Done.

pdgendt commented on drivers/flash/flash_mcux_flexspi_mx25um51345g.c:
```diff
@@ -0,0 +1,550 @@
+/*
+ * Copyright 2020 NXP
```
Nit, we are 2021

pdgendt commented on drivers/flash/flash_mcux_flexspi_mx25um51345g.c:
```diff
@@ -0,0 +1,550 @@
+}
+
+static int flash_flexspi_nor_erase_sector(const struct device *dev,
+	off_t offset)
```
Nit, inconsistent indentation.

pdgendt commented on dts/bindings/mtd/nxp,imx-flexspi-mx25um51345g.yaml:
```diff
@@ -0,0 +1,8 @@
+# Copyright 2020 NXP
```
Nit, also 2021

(Should have added the same comment here, sorry)

galak commented:
Seeing a build failure on `samples/boards/stm32/power_mgmt/blinky/sample.boards.stm32.power_mgmt.blinky` which I've asked @erwango to look into.
